nuttx-patches/priority_restoration_fix.patch

```diff
@@ -0,0 +1,448 @@
+ 
+ #if CONFIG_SEM_PREALLOCHOLDERS > 0
+   pholder = g_freeholders;
+-  if (pholder)
++  if (pholder != NULL)
+     {
+       /* Remove the holder from the free list an put it into the semaphore's
+        * holder list
+@@ -108,18 +108,24 @@ static inline FAR struct semholder_s *sem_allocholder(sem_t *sem)
+       pholder->counts  = 0;
+     }
+ #else
+-  if (!sem->holder.htcb)
++  if (sem->holder[0].htcb == NULL)
+     {
+-      pholder          = &sem->holder;
++      pholder          = &sem->holder[0];
++      pholder->counts  = 0;
++    }
++  else if (sem->holder[1].htcb == NULL)
++    {
++      pholder          = &sem->holder[1];
+       pholder->counts  = 0;
+     }
+ #endif
+   else
+     {
+       serr("ERROR: Insufficient pre-allocated holders\n");
+-      pholder = NULL;
++      pholder          = NULL;
+     }
+ 
++  DEBUGASSERT(pholder != NULL);
+   return pholder;
+ }
+ 
+@@ -132,16 +138,29 @@ static FAR struct semholder_s *sem_findholder(sem_t *sem,
+ {
+   FAR struct semholder_s *pholder;
+ 
++#if CONFIG_SEM_PREALLOCHOLDERS > 0
+   /* Try to find the holder in the list of holders associated with this
+    * semaphore
+    */
+ 
+-#if CONFIG_SEM_PREALLOCHOLDERS > 0
+-  for (pholder = sem->hhead; pholder; pholder = pholder->flink)
++  for (pholder = sem->hhead; pholder != NULL; pholder = pholder->flink)
++    {
++      if (pholder->htcb == htcb)
++        {
++          /* Got it! */
++
++          return pholder;
++        }
++    }
+ #else
+-  pholder = &sem->holder;
+-#endif
++  int i;
++  pholder = NULL;
++
++  /* We have two hard-allocated holder structuse in sem_t */
++
++  for (i = 0; i < 2; i++)
+     {
++      pholder = &sem->holder[i];
+       if (pholder->htcb == htcb)
+         {
+           /* Got it! */
+@@ -149,6 +168,7 @@ static FAR struct semholder_s *sem_findholder(sem_t *sem,
+           return pholder;
+         }
+     }
++#endif
+ 
+   /* The holder does not appear in the list */
+ 
+@@ -194,11 +214,11 @@ static inline void sem_freeholder(sem_t *sem, FAR struct semholder_s *pholder)
+        curr && curr != pholder;
+        prev = curr, curr = curr->flink);
+ 
+-  if (curr)
++  if (curr != NULL)
+     {
+       /* Remove the holder from the list */
+ 
+-      if (prev)
++      if (prev != NULL)
+         {
+           prev->flink = pholder->flink;
+         }
+@@ -216,6 +236,24 @@ static inline void sem_freeholder(sem_t *sem, FAR struct semholder_s *pholder)
+ }
+ 
+ /****************************************************************************
++ * Name: sem_findandfreeholder
++ ****************************************************************************/
++
++static inline void sem_findandfreeholder(sem_t *sem, FAR struct tcb_s *htcb)
++{
++  FAR struct semholder_s *pholder = sem_findholder(sem, htcb);
++
++  /* When no more counts are held, remove the holder from the list.  The
++   * count was decremented in sem_releaseholder.
++   */
++
++  if (pholder != NULL && pholder->counts <= 0)
```

Comment:
The `<= 0` confused me but I guess it's for the case it got decremented too far? Is `counts` signed?